### ISSUE TITLE
Route handler functions have `this`/context overridden

### ISSUE THREAD

**Issue by hoxxep** [restify/node-restify#981]
This may be intended behaviour, but it removes the possibility of using [IoC](https://github.com/jaredhanson/electrolyte) with ES6 classes, or makes it a lot more difficult, because class variables cannot be accessed since the `this` keyword is overridden.

In node v4.0.0+ running the following example and accessing `localhost:8080/action` would be expected to work, but instead it throws an error because it cannot access `this.SomethingModel` since the context/`this` is overridden. Is there an option to disable overriding the context on route handlers, or could such an option be added?

For the next major version, I'd request that the information contained in the overridden context/`this` be contained somewhere within `req` instead. Obviously there are some minor backwards compatibility issues, but it should be simple enough for users to fix when upgrading to the next major version of restify, and future-proofs restify a little better.

```javascript
'use strict';

var restify = require('restify');

class SomethingModel {
  action() {
    return Promise.resolve('Hello, World.');
  }
}

class SomethingController {
  constructor(SomethingModel) {
    this.SomethingModel = SomethingModel;
  }

  action(req, res, next) {
    // overridden with object
    console.log(this);

    // can't access this.SomethingModel as `this` is overridden
    // so any class variables are inaccessible!?
    this.SomethingModel.action()
      .then(function (result) {
        res.send(result);
        next();
      })
      .catch(function (error) {
        next(error);
      });
  }
}

class SomethingRoutes {
  constructor(SomethingController) {
    this.SomethingController = SomethingController;
  }

  routes(api) {
    api.get('/action', this.SomethingController.action);
  }
}

// electrolyte is being used for IoC and dependency injection
// but to KISS the example it's done manually here
var model = new SomethingModel();
var controller = new SomethingController(model);
var routes = new SomethingRoutes(controller);

var api = restify.createServer({
  name: 'Controller as ES6 Class Issue',
  version: '1.0.0'
});

routes.routes(api);

api.listen(8080, function () {
  console.log('App is running on localhost:8080');
});
```

**Comment by micahr** [comment]
Have you tried wrapping the action call with an anonymous function to close over the `this` context? 

For example:
```javascript
api.get('/action', function(req, res, next) {
    this.SomethingController.action(req, res, next);
});
```

This should guarantee that the `action` method is called with the correct `this` context.

**Comment by hoxxep** [comment]
That simply moves the context issue from the controller into the route, because the `this` keyword is still being overridden, meaning `this.SomethingController` can't be accessed. My current workaround is to hold the `this` keyword in a (Python-style named) `self` variable which is outside the class, and wrap the entire class in a constructor function like the demo below.

Thanks for the workaround suggestion though. Since I think class-based JS is the direction JavaScript is heading, due to the cleaner syntax, would you [and the other core developers] be averse to either:

- an option to disable overriding the context on routes (a slightly ugly but backwards compatible change)
- moving the information from the overridden context into the `req` object, and stop overriding the context altogether (a cleaner but breaking change)

I wouldn't mind writing the change and submitting a PR for v5.0 if one of these changes is deemed suitable by the core team.

```javascript
function SomethingRoutes(SomethingController) {
  var self = null;

  class SomethingRoutes {
    constructor(SomethingController) {
      self = this;
      self.SomethingController = SomethingController;
    }

    routes(api) {
      api.get('/action', function (req, res, next) {
        self.SomethingController.action(req, res, next);
      });
    }
  }

  return new SomethingRoutes(SomethingController);
}

module.exports = SomethingRoutes;
module.exports['@requires'] = ['SomethingController'];
```

**Comment by hoxxep** [comment]
I've now got a cleaner solution using arrow functions so that `this` is lexical, but it's still a bit of a hack and not as clean as it should be:

```javascript
class SomethingRoutes {
  constructor(SomethingController) {
    this.SomethingController = SomethingController;
  }

  routes(api) {
    api.get('/action', (req, res, next) => this.SomethingController.action(req, res, next));
    api.get('/action2', (req, res, next) => this.SomethingController.action2(req, res, next));
  }
}

// electrolyte IoC syntax
module.exports = SomethingRoutes;
module.exports['@requires'] = ['SomethingController'];
```

**Comment by DonutEspresso** [comment]
I can't see any reason why the handlers need to be bound to the server object. All tests are passing when bound to the global object. That said, If you need a specific context you should bind it manually in your SomethingController function. This is JS, functions are first class, there's no guarantee that context isn't being bound and re-bound again as the function is passed around. Even if ES6 classes take off, `call` and `apply` aren't ever going to go away.

For your example to work, it would require that the entire function call stack avoids binding the context in any way. Even if restify were to not do that (which it technically could), there's no guarantee that other utilities or libraries that restify depends on do the same. Case in point: domains. restify currently uses domains to execute handler functions, which are automatically bound to the domain it's running in. 

TL;DR, bind your context if you need it, it's the only way to guarantee the behavior you want.

**Comment by hoxxep** [comment]
That is true. I guess I'll need to wait for [javascript decorators](https://github.com/wycats/javascript-decorators) in the hope that [`@autobind`](https://github.com/andreypopp/autobind-decorator) is supported for a cleaner solution to this context binding issue. I'll close this issue since a workaround has been found. Thanks for the help :+1:
